src/HealthChecks.SmbCifs/SmbCifsStorageHealthCheck.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using SharpCifs.Smb;
+
+namespace HealthChecks.SmbCifs
+{
+    public class SmbCifsStorageHealthCheck : IHealthCheck
+    {
+        private readonly object _cifsOptions;
+
+        private NtlmPasswordAuthentication _auth;
+
+        public SmbCifsStorageHealthCheck(SmbCifsBasicOptions cifsOptions)
+        {
+            _cifsOptions = cifsOptions ?? throw new ArgumentNullException(nameof(cifsOptions));
+
+            if (_cifsOptions.GetType() != typeof(SmbCifsBasicOptions))
+                throw new InvalidCastException($"Invalid cast for object {nameof(cifsOptions)} to SmbCifsBasicOptions");
```

Comment:
InvalidCastException -> ArgumentException

Reply:
But I don’t understand why this check is necessary at all.

Reply:
Yup it's only a dumb check but was more for of any error on tests to break.
What would you suggest?

Reply:
What did you want to achieve?

Reply:
Well this 2 classes are just for are just for 2 different options to pass to the contri for of the service , my initial thought was to use one class only, but then I thing on the construction of setup of healthcheck that no one could distinguish between the 2 constructors parameters as they are different , so this is why made the the 2 concrete classes and then I thought using cast but before I wanted to make sure the cast was really ok on each ctor. option. Does this make sense to you?

Reply:
You may inherit `SmbCifsBasicOptions` and `SmbCifsExtendedOptions` from `SmbCifsOptions` with `Hostname`, `Domain`, `Username` properties in it. Then use `private readonly SmbCifsOptions_cifsOptions;` not `object`. In addition, hard type checks will not work in the case of inheritance. Consider this variant: `if (_cifsOptions is SmbCifsBasicOptions basic)` or `if (_cifsOptions is SmbCifsExtendedOptions extended)`

Reply:
Ok the word "is" was introduced in c# 7.1 so I guess I am still doing some old school code 😁. Of course the cast will not work with inehritance so I did not even though in that approach , but as soon as I have some time I will change and commit those changes and recheck unit tests. Anyway I hope I am contributing , that is the point right? And thanks for all your reviews and tips.

Reply:
> Anyway I hope I am contributing , that is the point right?

This repository is for collecting different health checks. The more of them the better. I think so.